src/Microsoft.ML.Data/Transforms/SelectColumnsTransform.cs

```diff
@@ -0,0 +1,632 @@
+                                                                             bool keepHidden,
+                                                                             Schema inputSchema)
+            {
+                var outputToInputMapping = new List<int>();
+                var columnCount = inputSchema.ColumnCount;
+                int outputIdx = 0;
+
+                for (int colIdx = 0; colIdx < columnCount; ++colIdx)
+                {
+                    if (!keepHidden && inputSchema.IsHidden(colIdx))
+                        continue;
+
+                    var columnName = inputSchema[colIdx].Name;
+                    if (selectedColumns.Contains(columnName))
+                    {
+                        outputToInputMapping.Add(colIdx);
+                        outputIdx++;
+                    }
+                }
+
+                return outputToInputMapping.ToArray();
+            }
+
+            private static Schema GenerateOutputSchema(IEnumerable<string> selectedColumns,
+                                                        bool keepHidden,
+                                                        Schema inputSchema)
+            {
+                var schemaColumns = new List<Schema.Column>();
+                var columnCount = inputSchema.ColumnCount;
+                for (int colIdx = 0; colIdx < columnCount; ++colIdx)
+                {
+                    if (!keepHidden && inputSchema.IsHidden(colIdx))
+                        continue;
+
+                    var column = inputSchema[colIdx];
+                    if (selectedColumns.Contains(column.Name))
+                    {
+                        schemaColumns.Add(column);
+                    }
+                }
+                return new Schema(schemaColumns);
+            }
+        }
+
+        private sealed class Row : IRow
+        {
+            private readonly Mapper _mapper;
+            private readonly IRow _input;
+            public Row(IRow input, Mapper mapper, Schema outputSchema)
+            {
+                _mapper = mapper;
+                _input = input;
+            }
+
+            public long Position => _input.Position;
+
+            public long Batch => _input.Batch;
+
+            public ISchema Schema => _mapper.Schema;
+
+            public ValueGetter<TValue> GetGetter<TValue>(int col)
+            {
+                int index = _mapper[col];
+                return _input.GetGetter<TValue>(index);
+            }
+
+            public ValueGetter<UInt128> GetIdGetter()
+                => _input.GetIdGetter();
+
+            public bool IsColumnActive(int col)
+            {
+                return true;
+            }
+        }
+
+        private sealed class SelectColumnsDataTransform : IDataTransform, IRowToRowMapper
+        {
+            private readonly Mapper _mapper;
+            private readonly IHostEnvironment _host;
```

Comment:
>_host [](start = 46, length = 5)

i would call it _env if it's IHostEnvironment.

Reply:
Updated the file as well.

---
In reply to: [225346508](https://github.com/dotnet/machinelearning/pull/1269#discussion_r225346508) [](ancestors = 225346508)